src/main/java/com/connexta/store/service/impl/StoreServiceImpl.java

```diff
@@ -25,42 +28,71 @@
   @NotNull private final IndexDatasetClient indexDatasetClient;
 
   @Override
-  public @NotNull URI createDataset(
-      @NotNull @Min(1L) @Max(10737418240L) Long fileSize,
-      @NotBlank String mediaType,
-      @NotBlank String fileName,
-      @NotNull InputStream fileInputStream)
-      throws CreateDatasetException, URISyntaxException {
-    final String key = UUID.randomUUID().toString().replace("-", "");
-    storageAdaptor.store(fileSize, mediaType, fileName, fileInputStream, key);
-    return new URI(retrieveEndpoint + key);
+  @NotNull
+  public URI createDataset(
+      @NotNull @Min(1L) @Max(10737418240L) final Long fileSize,
+      @NotBlank final String mediaType,
+      @NotBlank final String fileName,
+      @NotNull final InputStream fileInputStream)
+      throws StoreException, URISyntaxException {
+    final String datasetId = UUID.randomUUID().toString().replace("-", "");
+    fileStorageAdaptor.store(
+        fileSize, mediaType, fileInputStream, datasetId, Map.of(FILE_NAME_METADATA_KEY, fileName));
+    return new URI(retrieveEndpoint + datasetId);
   }
 
   @Override
-  public @NotNull RetrieveResponse retrieveFile(
+  public @NotNull FileRetrieveResponse retrieveFile(
       @Pattern(regexp = "^[0-9a-zA-Z]+$") @Size(min = 32, max = 32) final String datasetId)
       throws RetrieveException {
-    return storageAdaptor.retrieve(datasetId);
+    final StorageAdaptorRetrieveResponse storageAdaptorRetrieveResponse =
+        fileStorageAdaptor.retrieve(datasetId);
+
+    final String fileName =
+        storageAdaptorRetrieveResponse.getMetadata().get(FILE_NAME_METADATA_KEY);
+    if (StringUtils.isEmpty(fileName)) {
+      throw new RetrieveException(
+          String.format(
+              "Expected S3 object to have a non-null metadata value for %s",
+              FILE_NAME_METADATA_KEY));
+    }
+
+    return new FileRetrieveResponse(
+        storageAdaptorRetrieveResponse.getMediaType(),
+        storageAdaptorRetrieveResponse.getInputStream(),
+        fileName);
+  }
+
+  @Override
+  public @NotNull InputStream retrieveIrm(@NotBlank String datasetId) throws RetrieveException {
+    return irmStorageAdaptor.retrieve(datasetId).getInputStream();
   }
 
   @Override
-  public void addMetadata(
+  public void addIrm(
       @NotNull final InputStream irmInputStream,
       @NotNull @Min(1L) @Max(10737418240L) final long fileSize,
       @Pattern(regexp = "^[0-9a-zA-Z]+$") @Size(min = 32, max = 32) final String datasetId)
-      throws IndexMetadataException {
+      throws IndexDatasetException, URISyntaxException {
     // TODO check that the dataset exists
-    // TODO store IRM
-    indexDatasetClient.indexDataset(irmInputStream, fileSize, datasetId);
+    irmStorageAdaptor.store(
+        fileSize, StoreController.IRM_MEDIA_TYPE_VALUE, irmInputStream, datasetId, Map.of());
+    log.info("Successfully stored IRM for datasetId={}", datasetId);
+    indexDatasetClient.indexDataset(
```

Comment:
At some point in time we should rethink where we call `indexDatasetClient`. It is a method on StoreManager, but it isn't really being stored in the the same way a File is stored. The Metadata is being parsed and made searchable. Some of it (maybe all?) will end up persisted in s search engine. 

I think the right place to call `indexDatasetClient` is from the controllers. Controllers are a natural place to to use an HTTP client. Controllers take data and compose it to create domain objects (File, Metacard, ...), then pass those objects into the service layer.

Reply:
This comment might be outdated. We have been following the structure of each Controller has a 1 corresponding Service. IndexDatasetClient is only used in the StoreServiceImpl. I think Clients should only deal with the REST/spring stuff and not contain any app logic.